### PR TITLE
doc/doxygen: add BUILD_IN_DOCKER extended doc

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

How to correctly pass variables to DOCKER is missing some doc, so this adds some more doc on `BUILD_IN_DOCKER`

### Testing procedure

Check the doc.

### Issues/PRs references

#17770

### CODE REVIEW

File: doc/doxygen/src/build-in-docker.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Build In Docker                                       {#build-in-docker}
+
+## Targets ran in Docker: `DOCKER_MAKECMDGOALS_POSSIBLE`
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:3]:
```suggestion
## Targets ran in Docker: DOCKER_MAKECMDGOALS_POSSIBLE
```

File: doc/doxygen/src/build-in-docker.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+Currently only build related targets are ran in the docker container, the exact
+list is under `DOCKER_MAKECMDGOALS_POSSIBLE` variable.
+
+## Environment Variables: `DOCKER_ENV_VARS`
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:8]:
```suggestion
## Environment Variables: DOCKER_ENV_VARS
```

File: doc/doxygen/src/build-in-docker.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+docker run --rm -t -u "$(id -u)" \
+    ...
+    -e 'BEER_TYPE=imperial stout' \
+    -w '/data/riotbuild/riotbase/examples/hello-world/' \
+    'riot/riotbuild:latest' make
+```
+
+Your application Makefile can also extend `DOCKER_ENV_VARS`.
+
+### Directly Define Environment Variables: `DOCKER_ENVIRONMENT_CMDLINE`
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:35]:
```suggestion
### Directly Define Environment Variables: DOCKER_ENVIRONMENT_CMDLINE
```

File: doc/doxygen/src/build-in-docker.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+```
+DOCKER_ENVIRONMENT_CMDLINE='-e BEER_TYPE="imperial stout"' BUILD_IN_DOCKER=1 make -C examples/hello-world/
+docker run --rm -t -u "$(id -u)" \
+    ...
+    -e 'BEER_TYPE=imperial stout' \
+    -w '/data/riotbuild/riotbase/examples/hello-world/' \
+    'riot/riotbuild:latest' make
+```
+
+## Command line Variables: `DOCKER_OVERRIDE_CMDLINE`
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:51]:
```suggestion
## Command line Variables: DOCKER_OVERRIDE_CMDLINE
```

File: doc/doxygen/src/build-in-docker.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+```shell
+DOCKER_OVERRIDE_CMDLINE="BEER_TYPE='imperial stout'" BUILD_IN_DOCKER=1 make -C examples/hello-world/ RIOT_CI_BUILD=1
+Launching build container using image "riot/riotbuild:latest".
+sudo docker run --rm -t -u "$(id -u)" \
+    ...
+    -w '/data/riotbuild/riotbase/examples/hello-world/' \
+    'riot/riotbuild:latest' make  BEER_TYPE='imperial stout' 'RIOT_CI_BUILD=1'
+```
+
+### Redefined or Overridden Variables: `DOCKER_ENV_VARS_ALWAYS`
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:71]:
```suggestion
### Redefined or Overridden Variables: DOCKER_ENV_VARS_ALWAYS
```

File: doc/doxygen/src/build-in-docker.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+to `file` and there is no way of detecting in Make how it was set.
+
+If this happens after `docker.ink.mk` this is not an issue, but for all variables
+susceptible to be defined in the application `Makefile` this is indeed the case.
+
+A subset of these variables, namely variables relating to dependency resolution
+are therefore unconditionally passed to docker. The complete list can be found
+under `DOCKER_ENV_VARS_ALWAYS`.
+
+#### `CFLAGS`
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:84]:
```suggestion
#### CFLAGS
```

File: doc/doxygen/src/build-in-docker.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    -w '/data/riotbuild/riotbase/examples/hello-world/' \
+    'riot/riotbuild:latest' make
+```
+
+Your application Makefile can also extend `DOCKER_ENV_VARS`.
+
+### Directly Define Environment Variables: `DOCKER_ENVIRONMENT_CMDLINE`
+
+`DOCKER_ENVIRONMENT_CMDLINE` can be used to add variables directly to the environment
+but will need to be pDOCKER_ENV_VARS=BEER_TYPE refixed with `-e` (see [option-summary[(https://www.gnu.org/software/make/manual/html_node/Options-Summary.html)).
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:38]:
```suggestion
but will need to be pDOCKER_ENV_VARS=BEER_TYPE refixed with `-e` (see [option-summary](https://www.gnu.org/software/make/manual/html_node/Options-Summary.html)).
```

File: doc/doxygen/src/build-in-docker.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    'riot/riotbuild:latest' make
+```
+
+## Command line Variables: `DOCKER_OVERRIDE_CMDLINE`
+
+Command line arguments are variables that are passed after `make` e.g.
+`make target FOO=bar`, but different to environment variables a variable defined
+through the command line will take precedence over all assignments of `FOO` within
+the makefile (same effect as adding `-e` for environment variables, see
+[option-summary[(https://www.gnu.org/software/make/manual/html_node/Options-Summary.html) for more details.
```
Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:57]:
```suggestion
[option-summary](https://www.gnu.org/software/make/manual/html_node/Options-Summary.html) for more details.
```

Comment (fjmolinas) [inline doc/doxygen/src/build-in-docker.md:57]: Hmm links could also be moved to the end of the file instead.

**Review by fjmolinas — COMMENTED** [review]
Some things to fix found in the rendering, btw is the location ok?

image

**Review by chrysn — COMMENTED** [review]
Some fixes and enhancements (hope this does get through, I started getting errors while editing...)

File: doc/doxygen/src/build-in-docker.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+e.g: if wanting to override STRING_WITH_SPACES
+
+```
+# normal build
+CFLAGS=-DSTRING_WITH_SPACES='\"with space\" make
+# in docker
+DOCKER_ENVIRONMENT_CMDLINE='-e CFLAGS=-DSTRING_WITH_SPACES=\'\\\"with\ space\\\"\'' \
+  BUILD_IN_DOCKER=1 make
+```
+
```
Comment (chrysn) [inline doc/doxygen/src/build-in-docker.md:99]:
```suggestion

Alternatively, it is often easier to define the CFLAGS in the Makefile
(which gets evaluated inside the Docker image again),
conditional on a less complex environment variable
that gets added to `DOCKER_ENV_VARS` in the Makefile.
```

File: doc/doxygen/src/build-in-docker.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
+# Build In Docker                                       {#build-in-docker}
+
```
[the remark this answers is not in the document]
Comment (fjmolinas) [reply doc/doxygen/src/build-in-docker.md:2]: There is also some documentation on docker in the `getting-started` page, maybe I should move all the doc there here, and add a reference.

**Comment by fjmolinas** [comment]
> Some fixes and enhancements (hope this does get through, I started getting errors while editing...)

BTW does this PR answer the questions you had?

**Comment by chrysn** [comment]
> btw is the location ok

I think it'd be better right below "getting started", but then again the whole list might be due for an overhaul (which is larger than this issue, and I don't have any good concrete ideas yet as to what it should look like).

**Review by chrysn — APPROVED** [review]
Thanks, answers my questions; I think this is ready.

**Comment by fjmolinas** [comment]
Thanks for the review!

**Pull request merged**